### PR TITLE
Load balance statefulset DNS response

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/31064

Code is borrowed from coredns

### CODE REVIEW

**Review by hzxuzhonghu — APPROVED** [review]

**Review by adiprerepa — APPROVED** [review]

File: pilot/pkg/dns/dns.go [hunk 1]
```diff
@@ -206,10 +206,75 @@ func (h *LocalDNSServer) ServeDNS(proxy *dnsProxy, w dns.ResponseWriter, req *dn
...
+	out = append(out, mx...)
+	return out
+}
+
+func roundRobinShuffle(records []dns.RR) {
+	switch l := len(records); l {
+	case 0, 1:
+		break
+	case 2:
+		log.Errorf("howardjohn: ID: %v", dns.Id())
```
Comment (ramaraochavali) [inline pilot/pkg/dns/dns.go:262]: nit: remove this log?

File: releasenotes/notes/dns-round-robin.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+apiVersion: release-notes/v2
+kind: feature
+area: networking
+issue: [31064]
```
Comment (shamsher31) [inline releasenotes/notes/dns-round-robin.yaml:4]: Not sure if [] works for issues. @brian-avery can you please confirm?
```suggestion
issue: 31064
```

Comment (howardjohn) [reply releasenotes/notes/dns-round-robin.yaml:4]: I had that and the test failed telling me to use a list

Comment (shamsher31) [reply releasenotes/notes/dns-round-robin.yaml:4]: Ok, so it should be.
```
issue:
  - 31064
```

**Review by shamsher31 — APPROVED** [review]

File: releasenotes/notes/dns-round-robin.yaml (hunk 2 above)
Comment (howardjohn) [reply releasenotes/notes/dns-round-robin.yaml:4]: that is equivalent to what I have currently

**Comment by howardjohn** [comment]
hold to remove log

**Review by ramaraochavali — APPROVED** [review]

**Review by ericvn — APPROVED** [review]

**Comment by howardjohn** [comment]
I think this exposes a real bug/known limitation, which is why the tests fail. We are load balancing across multiple clusters now, including multi-network, which doesn't actually work.

**Comment by howardjohn** [comment]
Ok, so what is happening. Say we have cluster A and B, and both deploy a headless service with instances a in A and b in B.

Without DNS interception, `curl svc` in A will always return ip-a. `curl svc` in B will always return ip-b. Result: traffic works, but is always kept in the same cluster
With DNS interception, before this PR: `curl svc` in A will always return [ip-a, ip-b]. `curl svc` B will always return [ip-b, ip-a]. Result: the tests pass since we only use the first IP in the response, so we always send to the local cluster. Load balancing in the same cluster is broken for the same rason
With DNS, with this PR: `curl svc` in either cluster will always return [ip-a, ip-b] or  [ip-b, ip-a]. Result: tests break since we send cross-network traffic through headless service. But local cluster load balancing works.

So it sounds like we need the randomization of DNS like in this change, but Istiod needs to change to only send IPs from the local cluster. Thoughts @ramaraochavali @hzxuzhonghu @nmittler ?

**Comment by nmittler** [comment]
@howardjohn Yes, I think that's right

**Comment by ramaraochavali** [comment]
@howardjohn That is an awesome find.

**Pull request merged**